File: apps/game-ios/Sources/ALUI/Shop/Meter.swift
import QuartzCore
import SwiftUI

import ALCore

/* -------------------------------------------------------------------------- */
/* `src/shop/Meter.tsx` — SavingsMeter: "how close am I to affording this?"     */
/* without any arithmetic.                                                     */
/*                                                                             */
/* ```tsx                                                                      */
/* const MIN_FILL = 0.07;                                                      */
/*                                                                             */
/* function fillRatio(balance: number, cost: number): number {                 */
/*   if (balance <= 0 || cost <= 0) return 0;                                  */
/*   return Math.max(MIN_FILL, Math.min(1, balance / cost));                   */
/* }                                                                           */
/*                                                                             */
/* export function SavingsMeter({ cost, balance, since, height = 10 }) {       */
/*   const shownRef = useRef(fillRatio(since, cost)); // ratio on screen        */
/*   const ratio = fillRatio(balance, cost);                                    */
/*   useEffect(() => {                                                          */
/*     const from = shownRef.current;                                           */
/*     shownRef.current = ratio;                                                */
/*     if (from === ratio) return;                                              */
/*     meterFill(fillRef.current, from * 100, ratio * 100);                     */
/*     if (ratio > from) tipSparkle(tipRef.current);                            */
/*   }, [ratio]);                                                               */
/*   … role="img" aria-label={`${balance} étoiles sur ${cost}`}                 */
/*     track  rounded-full, height, rgba(90,58,30,0.14)                         */
/*     fill   width: `${ratio * 100}%`, linear-gradient(90deg,#FFC107,#FFD54F)  */
/*     tip    ✨ absolute left: `${ratio * 100}%`, top: 50%, opacity-0           */
/* }                                                                            */
/* ```                                                                          */
/*                                                                             */
/* A fat gold bar filling toward the price: a six-year-old reads fills, not      */
/* subtraction. On mount it sweeps from the balance the child LAST SAW in the    */
/* shop (`since`) to today's, so stars earned between visits read as motion      */
/* plus a sparkle rather than an invisibly small static delta. A non-empty       */
/* wallet always shows a sliver (`MIN_FILL`), so "I have some" is always         */
/* visible.                                                                     */
/*                                                                             */
/* INVARIANT 9 — nothing here persists, accumulates or computes a balance.       */
/* `balance` and `since` are handed in by the shop, which reads them from        */
/* `ProfileStore`; this file only divides one number by another to get a width.  */
/* -------------------------------------------------------------------------- */

// MARK: - The arithmetic (pure, host-tested)

public enum SavingsMeterMetrics {
    /// `MIN_FILL` — never render progress smaller than this once the wallet is
    /// non-empty.
    public static let minFill: Double = 0.07

    /// `height = 10` — the TSX default. "Keep it fat: thin bars hide small
    /// progress."
    public static let defaultHeight: CGFloat = 10

    /// `rgba(90,58,30,0.14)` — the track. 90/58/30 is `#5A3A1E`, i.e. the ink
    /// brown at 14 %.
    public static let trackOpacity: Double = 0.14

    /// `text-sm` — the ✨ at the fill tip.
    public static let sparkleFontSize: CGFloat = Typography.Size.sm

    /// `meterFill` — `{ duration: 900, easing: "cubic-bezier(.2,.7,.3,1)" }`.
    public static let sweepDuration: Double = 0.9
    public static let sweepCurve: (Double, Double, Double, Double) = (0.2, 0.7, 0.3, 1)

    /// `tipSparkle` — `{ duration: 620, delay: 780, easing: "ease-out" }`,
    /// scale 0.2 → 1.4 (at 50 %) → 0.6, opacity 0 → 1 → 0. "lands as the fill
    /// arrives."
    public static let sparkleDuration: Double = 0.62
    public static let sparkleDelay: Double = 0.78
    public static let sparkleScales: [Double] = [0.2, 1.4, 0.6]
    public static let sparkleOpacities: [Double] = [0, 1, 0]
}

/// `fillRatio(balance, cost)` — 0…1, with the `MIN_FILL` sliver.
///
/// Both guards are the TSX's and both matter: a zero or negative balance is a
/// genuinely empty bar (no sliver), and a zero or negative cost would divide by
/// nothing.
public func savingsFillRatio(balance: Int, cost: Int) -> Double {
    if balance <= 0 || cost <= 0 { return 0 }
    return Swift.max(
        SavingsMeterMetrics.minFill,
        Swift.min(1, Double(balance) / Double(cost)))
}

/// What the mount effect decides: where the bar starts, where it lands, whether
/// it moves at all, and whether the tip sparkles.
///
/// Extracted so the rule — *sparkle only when the fill GREW* — is asserted on
/// the host rather than only visible on a device.
public struct SavingsMeterSweep: Equatable, Sendable {
    /// `from = shownRef.current`, i.e. `fillRatio(since, cost)` on mount.
    public let from: Double
    /// `ratio`, i.e. `fillRatio(balance, cost)`.
    public let to: Double

    public init(from: Double, to: Double) {
        self.from = from
        self.to = to
    }

    /// `if (from === ratio) return;` — no animation when nothing changed.
    public var animates: Bool { from != to }

    /// `if (ratio > from) tipSparkle(...)` — a shrinking bar (the child spent)
    /// still sweeps, silently.
    public var sparkles: Bool { to > from }
}

/// The mount sweep for a (balance, since, cost) triple.
public func savingsMeterSweep(balance: Int, since: Int, cost: Int) -> SavingsMeterSweep {
    SavingsMeterSweep(
        from: savingsFillRatio(balance: since, cost: cost),
        to: savingsFillRatio(balance: balance, cost: cost))
}

// MARK: - The view

/// `<SavingsMeter cost balance since height />`.
public struct SavingsMeter: View {
    private let cost: Int
    private let balance: Int
    private let since: Int
    private let height: CGFloat

    @Environment(\.alReduceMotion) private var reduceMotion

    /// The ratio currently ON SCREEN — the port of `shownRef`. Seeded from
    /// `since` in `init`, exactly as the ref is, so the first frame paints where
    /// the child left the bar and the sweep has somewhere to come from.
    @State private var shown: Double
    @State private var sparkleHandle = LayerHandle()

    /// - Parameters:
    ///   - cost: the price the bar fills toward.
    ///   - balance: the child's stars right now.
    ///   - since: the balance at the previous shop visit — where the fill
    ///     animates FROM.
    ///   - height: bar height in points. Keep it fat.
    public init(cost: Int, balance: Int, since: Int, height: CGFloat = SavingsMeterMetrics.defaultHeight) {
        self.cost = cost
        self.balance = balance
        self.since = since
        self.height = height
        _shown = State(initialValue: savingsFillRatio(balance: since, cost: cost))
    }

    /// `fillRatio(balance, cost)` — the width React writes into `style`.
    public var ratio: Double { savingsFillRatio(balance: balance, cost: cost) }

    /// `aria-label={`${balance} étoiles sur ${cost}`}` on a `role="img"`.
    public var accessibilityText: String { Copy.Shop.savings(balance, of: cost) }

    public var body: some View {
        GeometryReader { proxy in
            let width = proxy.size.width
            ZStack(alignment: .leading) {
                // `block w-full overflow-hidden rounded-full` + the 14 % track.
                Capsule()
                    .fill(Palette.ink.color.opacity(SavingsMeterMetrics.trackOpacity))

                // `block h-full rounded-full` + `linear-gradient(90deg,…)`.
                // Clipped by the track, so a full bar has square-ish shoulders
                // exactly as `overflow-hidden` gives on the web.
                Capsule()
                    .fill(Palette.savingsFill.gradient)
                    .frame(width: Swift.max(0, shown * width))

                // The sparkle is parked (invisible) AT the fill tip and flashed
                // by `tipSparkle`; the outer span is `overflow-visible` so it
                // may hang past the bar.
                LayerHost(handle: sparkleHandle) {
                    Text(verbatim: Copy.Shop.savingsSparkle)
                        .font(Typography.rounded(SavingsMeterMetrics.sparkleFontSize))
                        .opacity(0)
                }
                .allowsHitTesting(false)
                .accessibilityHidden(true)
                .position(x: ratio * width, y: height / 2)
            }
            .frame(height: height)
            .onAppear { sweep(to: ratio) }
            .onChange(of: ratio) { _, next in sweep(to: next) }
        }
        .frame(height: height)
        .accessibilityElement(children: .ignore)
        .accessibilityAddTraits(.isImage)
        .accessibilityLabel(Text(verbatim: accessibilityText))
    }

    /// `useEffect(..., [ratio])`: read where the bar stands, remember the new
    /// target, and animate the journey. Both animations are reduced-motion
    /// gated because `shop/anim.ts` gates them (`meterFill`/`tipSparkle` both
    /// start `if (!el || reducedMotion()) return`), and the bar still LANDS on
    /// the new value either way — a child with Reduce Motion on sees the right
    /// width, just not the trip.
    private func sweep(to next: Double) {
        let step = SavingsMeterSweep(from: shown, to: next)
        guard step.animates else { return }
        guard !reduceMotion.isReduced else {
            shown = next
            return
        }
        withAnimation(
            .timingCurve(
                SavingsMeterMetrics.sweepCurve.0,
                SavingsMeterMetrics.sweepCurve.1,
                SavingsMeterMetrics.sweepCurve.2,
                SavingsMeterMetrics.sweepCurve.3,
                duration: SavingsMeterMetrics.sweepDuration)
        ) {
            shown = next
        }
        if step.sparkles { MeterAnim.tipSparkle(sparkleHandle.layer) }
    }
}

// MARK: - The tip sparkle, as Core Animation

/// `tipSparkle` from `src/shop/anim.ts`, transform + opacity only, on the hosted
/// layer — off the render path (invariant 2), exactly like `Anim.press`.
///
/// Kept local to this file rather than added to `Anim`: `Anim` is the *shared*
/// interaction vocabulary (press / shake / pop / pulse) and the shop's sparkle
/// is one screen's ornament. If the shop package grows its own `anim.ts` port,
/// this is the function to fold into it.
///
/// macOS (D1): `LayerHost` is a no-op there, so `layer` is nil and this returns
/// having done nothing — the meter still renders and still lands on the right
/// width. The keyframes are asserted as data in the tests instead.
enum MeterAnim {

    @MainActor
    static func tipSparkle(_ layer: CALayer?) {
        guard let layer else { return }
        layer.add(sparkleAnimation(), forKey: "ALMeterTipSparkle")
    }

    /// `[{scale 0.2, opacity 0}, {scale 1.4, opacity 1, offset .5},
    ///   {scale 0.6, opacity 0}]`, 620 ms, 780 ms delay, ease-out.
    ///
    /// The web's `translate(-50%,-50%)` is not part of this: the layer is
    /// already centred on the fill tip by `.position(x:y:)`, so only the scale
    /// and the opacity are animated. No `fill` mode, so the model values (a
    /// resting opacity of 0) stand once the take ends — the sparkle vanishes.
    static func sparkleAnimation() -> CAAnimationGroup {
        let scale = keyframes(
            keyPath: "transform.scale",
            values: SavingsMeterMetrics.sparkleScales)
        let opacity = keyframes(
            keyPath: "opacity",
            values: SavingsMeterMetrics.sparkleOpacities)
        let group = CAAnimationGroup()
        group.animations = [scale, opacity]
        group.duration = SavingsMeterMetrics.sparkleDuration
        group.beginTime = CACurrentMediaTime() + SavingsMeterMetrics.sparkleDelay
        return group
    }

    /// WAAPI's `easing` applies per keyframe INTERVAL, not to the whole
    /// timeline — the same porting note `Anim` carries.
    static func keyframes(keyPath: String, values: [Double]) -> CAKeyframeAnimation {
        let animation = CAKeyframeAnimation(keyPath: keyPath)
        animation.values = values
        animation.keyTimes = [0, 0.5, 1].map { NSNumber(value: $0) }
        animation.timingFunctions = (0..<(values.count - 1)).map { _ in
            CAMediaTimingFunction(controlPoints: 0, 0, 0.58, 1)  // CSS ease-out
        }
        animation.duration = SavingsMeterMetrics.sparkleDuration
        return animation
    }
}

File: apps/game-android/ui/src/main/kotlin/fr/dappit/attrapelettres/ui/shop/Meter.kt
package fr.dappit.attrapelettres.ui.shop

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.AnimationVector1D
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.Stable
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import fr.dappit.attrapelettres.core.platform.ReduceMotionSource
import fr.dappit.attrapelettres.ui.design.Copy
import fr.dappit.attrapelettres.ui.design.Palette
import fr.dappit.attrapelettres.ui.design.Typography
import fr.dappit.attrapelettres.ui.design.fixedSp
import fr.dappit.attrapelettres.ui.interaction.Anim
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

// ---------------------------------------------------------------------------
// `src/shop/Meter.tsx` — SavingsMeter: « how close am I to affording this? »
// without any arithmetic.
//
//   const MIN_FILL = 0.07;
//
//   function fillRatio(balance: number, cost: number): number {
//     if (balance <= 0 || cost <= 0) return 0;
//     return Math.max(MIN_FILL, Math.min(1, balance / cost));
//   }
//
//   export function SavingsMeter({ cost, balance, since, height = 10 }) {
//     const shownRef = useRef(fillRatio(since, cost));   // ratio on screen
//     const ratio = fillRatio(balance, cost);
//     useEffect(() => {
//       const from = shownRef.current;
//       shownRef.current = ratio;
//       if (from === ratio) return;
//       meterFill(fillRef.current, from * 100, ratio * 100);
//       if (ratio > from) tipSparkle(tipRef.current);
//     }, [ratio]);
//     … role="img" aria-label={`${balance} étoiles sur ${cost}`}
//   }
//
// A fat gold bar filling toward the price: a six-year-old reads fills, not
// subtraction. On mount it sweeps from the balance the child LAST SAW in the
// shop (`since`) to today's, so stars earned between visits read as motion plus
// a sparkle rather than an invisibly small static delta. A non-empty wallet
// always shows a sliver (`MIN_FILL`), so « I have some » is always visible.
//
// INVARIANT 8 / INVARIANT 9 — nothing here persists, accumulates or computes a
// balance. `balance` and `since` are handed in by the shop, which reads them
// from `ProfileStore`; this file divides one number by another to get a width
// and does nothing else with either.
//
// INVARIANT 2 — the fill width is an `Animatable` read inside a `drawBehind`
// lambda, not `Modifier.width(animatedDp)`. A width is a LAYOUT property: as
// snapshot state it would re-measure the growth card sixty times a second. As a
// painted band it costs one re-draw of this leaf. That is the same trade A12
// records for `Tile`'s highlight ring, and it is why the bar is drawn rather
// than sized.
// ---------------------------------------------------------------------------

// --- The arithmetic (pure, host-tested) --------------------------------------

object SavingsMeterMetrics {

    /**
     * `MIN_FILL` — never render progress smaller than this once the wallet is
     * non-empty.
     */
    const val MIN_FILL: Double = 0.07

    /** `height = 10` — the TSX default. « Keep it fat: thin bars hide small progress. » */
    val DEFAULT_HEIGHT: Dp = 10.dp

    /**
     * `rgba(90,58,30,0.14)` — the track. 90/58/30 is `#5A3A1E`, i.e. the ink
     * brown at 14 %.
     */
    const val TRACK_OPACITY: Float = 0.14f

    /** `text-sm` — the ✨ at the fill tip. */
    val SPARKLE_SIZE: Dp = Typography.Size.sm

    /** `meterFill` — `{ duration: 900, easing: "cubic-bezier(.2,.7,.3,1)" }`. */
    const val SWEEP_DURATION_MS: Int = 900
    val SWEEP_CURVE = ShopAnim.SPARK_CURVE

    /**
     * `tipSparkle` — `{ duration: 620, delay: 780, easing: "ease-out" }` over
     * `scale 0.2 → 1.4 (offset .5) → 0.6` and `opacity 0 → 1 → 0`. « Lands as
     * the fill arrives. »
     */
    const val SPARKLE_DURATION_MS: Int = 620
    const val SPARKLE_DELAY_MS: Int = 780

    val SPARKLE_SCALE = MotionTrack(
        values = listOf(0.2f, 1.4f, 0.6f),
        keyTimes = listOf(0f, 0.5f, 1f),
        easing = Anim.EASE_OUT,
    )

    val SPARKLE_ALPHA = MotionTrack(
        values = listOf(0f, 1f, 0f),
        keyTimes = listOf(0f, 0.5f, 1f),
        easing = Anim.EASE_OUT,
    )
}

/**
 * `fillRatio(balance, cost)` — 0…1, with the `MIN_FILL` sliver.
 *
 * Both guards are the TSX's and both matter: a zero or negative balance is a
 * genuinely empty bar (no sliver), and a zero or negative cost would divide by
 * nothing.
 *
 * `Double`, not `Float`: a JS number is a double and the test compares against a
 * transcription of the TypeScript, where `7 / 100` must be exactly `0.07`.
 */
fun savingsFillRatio(balance: Int, cost: Int): Double {
    if (balance <= 0 || cost <= 0) return 0.0
    return maxOf(
        SavingsMeterMetrics.MIN_FILL,
        minOf(1.0, balance.toDouble() / cost.toDouble()),
    )
}

/**
 * What the mount effect decides: where the bar starts, where it lands, whether
 * it moves at all, and whether the tip sparkles.
 *
 * Extracted so the rule — *sparkle only when the fill GREW* — is asserted on the
 * host rather than only visible on a device.
 */
data class SavingsMeterSweep(val from: Double, val to: Double) {

    /** `if (from === ratio) return;` — no animation when nothing changed. */
    val animates: Boolean get() = from != to

    /**
     * `if (ratio > from) tipSparkle(...)` — a shrinking bar (the child spent)
     * still sweeps, silently. No sparkle for losing stars.
     */
    val sparkles: Boolean get() = to > from
}

/** The mount sweep for a (balance, since, cost) triple. */
fun savingsMeterSweep(balance: Int, since: Int, cost: Int): SavingsMeterSweep =
    SavingsMeterSweep(
        from = savingsFillRatio(since, cost),
        to = savingsFillRatio(balance, cost),
    )

// --- The two animations ------------------------------------------------------

/**
 * The bar's on-screen ratio (the port of `shownRef`) and the tip sparkle.
 *
 * Both are `Animatable`s and neither is snapshot state a composable body reads.
 * Both are gated, because `shop/anim.ts` gates `meterFill` and `tipSparkle`
 * itself — and the bar still LANDS on the new value either way: a child with the
 * setting on sees the right width, just not the trip.
 */
@Stable
class SavingsMeterMotion(
    private val scope: CoroutineScope,
    private val reduceMotion: ReduceMotionSource,
    initialRatio: Float,
) {
    /** The ratio currently ON SCREEN. Read inside `drawBehind` only. */
    val shown: Animatable<Float, AnimationVector1D> = Animatable(initialRatio)

    /** The sparkle's timeline, 0..1. Read inside `graphicsLayer` only. */
    val sparkle: Animatable<Float, AnimationVector1D> = Animatable(0f)

    private var sweepJob: Job? = null
    private var sparkleJob: Job? = null

    val gated: Boolean get() = !ShopAnim.shouldAnimate(reduceMotion)

    /**
     * `useEffect(..., [ratio])`: read where the bar stands, remember the new
     * target, and animate the journey.
     */
    fun sweep(to: Double) {
        val step = SavingsMeterSweep(shown.value.toDouble(), to)
        if (!step.animates) return
        sweepJob?.cancel()
        if (gated) {
            sweepJob = scope.launch { shown.snapTo(to.toFloat()) }
            return
        }
        sweepJob = scope.launch {
            shown.animateTo(
                targetValue = to.toFloat(),
                animationSpec = tween(
                    durationMillis = SavingsMeterMetrics.SWEEP_DURATION_MS,
                    easing = SavingsMeterMetrics.SWEEP_CURVE.toEasing(),
                ),
            )
        }
        if (step.sparkles) flash()
    }

    /**
     * `tipSparkle`. LINEAR progress on purpose: the authored ease-out is
     * per-INTERVAL and already lives in the two [MotionTrack]s.
     *
     * No `fill` mode in the TSX's `el.animate(…)`, so the resting opacity of 0
     * stands once the take ends — the sparkle vanishes rather than sticking,
     * which is what snapping back to 0 reproduces.
     */
    private fun flash() {
        sparkleJob?.cancel()
        sparkleJob = scope.launch {
            sparkle.snapTo(0f)
            delay(SavingsMeterMetrics.SPARKLE_DELAY_MS.toLong())
            sparkle.animateTo(
                targetValue = 1f,
                animationSpec = tween(
                    durationMillis = SavingsMeterMetrics.SPARKLE_DURATION_MS,
                    easing = LinearEasing,
                ),
            )
            sparkle.snapTo(0f)
        }
    }
}

// --- The view -----------------------------------------------------------------

/**
 * `<SavingsMeter cost balance since height />`.
 *
 * @param cost the price the bar fills toward.
 * @param balance the child's stars right now.
 * @param since the balance at the previous shop visit — where the fill animates
 *   FROM. Seeded into [SavingsMeterMotion] at construction, exactly as the TSX
 *   seeds `shownRef`, so the first frame paints where the child left the bar and
 *   the sweep has somewhere to come from.
 * @param height bar height. Keep it fat.
 */
@Composable
fun SavingsMeter(
    cost: Int,
    balance: Int,
    since: Int,
    reduceMotion: ReduceMotionSource,
    modifier: Modifier = Modifier,
    height: Dp = SavingsMeterMetrics.DEFAULT_HEIGHT,
) {
    val ratio = savingsFillRatio(balance, cost)
    val scope = rememberCoroutineScope()
    val fontScale = LocalDensity.current.fontScale
    // Keyed on `cost` and NOT on `balance`: re-seeding on every balance change
    // would make the bar jump instead of sweep, which is the whole point of the
    // `since` prop.
    val motion = remember(scope, reduceMotion, cost) {
        SavingsMeterMotion(scope, reduceMotion, savingsFillRatio(since, cost).toFloat())
    }

    LaunchedEffect(motion, ratio) { motion.sweep(ratio) }

    // The track's width in px, for placing the sparkle at the fill tip. Written
    // on layout only — never per frame.
    val widthPx = remember { mutableFloatStateOf(0f) }

    val trackColor = Palette.ink.color.copy(alpha = SavingsMeterMetrics.TRACK_OPACITY)

    Box(
        modifier
            .fillMaxWidth()
            .height(height)
            .onSizeChanged { widthPx.floatValue = it.width.toFloat() }
            // `role="img"` + `aria-label={`${balance} étoiles sur ${cost}`}`.
            // Note « étoiles » is NOT pluralised by the TSX, so 1 star still
            // says « étoiles »; ported as authored.
            .semantics(mergeDescendants = true) {
                role = Role.Image
                contentDescription = Copy.Shop.savings(balance, cost)
            }
            .drawBehind {
                val radius = CornerRadius(size.height / 2f)
                // `block w-full overflow-hidden rounded-full` + the 14 % track.
                drawRoundRect(color = trackColor, cornerRadius = radius)
                // THE DEFERRED READ. Draw phase, so a frame of the sweep costs
                // one re-draw of this leaf and zero recompositions.
                val filled = (motion.shown.value * size.width).coerceIn(0f, size.width)
                if (filled > 0f) {
                    val band = Size(filled, size.height)
                    // `linear-gradient(90deg,#FFC107,#FFD54F)` across the FILL,
                    // not across the track — the gradient travels with the bar.
                    drawRoundRect(
                        brush = Palette.savingsFill.brush(band),
                        size = band,
                        cornerRadius = radius,
                    )
                }
            },
    ) {
        // « Sparkle parked (invisible) at the fill tip; tipSparkle() flashes
        // it. » `overflow-visible` on the outer span is why it may hang past the
        // bar, which a Compose `Box` allows by default.
        BasicText(
            text = Copy.Shop.SAVINGS_SPARKLE,
            style = TextStyle(
                fontSize = fixedSp(SavingsMeterMetrics.SPARKLE_SIZE.value, fontScale),
                fontFamily = Typography.appFamily,
            ),
            modifier = Modifier.graphicsLayer {
                val t = motion.sparkle.value
                val s = SavingsMeterMetrics.SPARKLE_SCALE.at(t)
                scaleX = s
                scaleY = s
                alpha = SavingsMeterMetrics.SPARKLE_ALPHA.at(t)
                // `left: ${ratio * 100}%; top: 50%` plus the web's
                // `translate(-50%,-50%)`.
                // `size` inside this lambda is the SPARKLE's own box; the bar's
                // height comes from the authored parameter, converted here
                // because a `GraphicsLayerScope` is a `Density`.
                translationX = ratio.toFloat() * widthPx.floatValue - size.width / 2f
                translationY = height.toPx() / 2f - size.height / 2f
            },
        )
    }
}
